[Site Isolation] Pointer Lock

stwrt opened the pull request:
#### d1dcb5ad565b1c852043d169b7937535b19b2c96
<pre>
[Site Isolation] Pointer Lock
<a href="https://bugs.webkit.org/show_bug.cgi?id=293767">https://bugs.webkit.org/show_bug.cgi?id=293767</a>
<a href="https://rdar.apple.com/117910429">rdar://117910429</a>

Reviewed by Alex Christensen.

Implement Pointer Lock to work under site isolation.

WK1 change: When disabling pointer lock options a pointerlockerror event will be thrown now
in place of pointerlockchange.

* LayoutTests/http/tests/site-isolation/pointer-lock-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/pointer-lock.html: Added.
* LayoutTests/http/tests/site-isolation/resources/pointer-lock.html: Added.
* LayoutTests/platform/ios/http/tests/site-isolation/pointer-lock-expected.txt: Added.
* LayoutTests/platform/mac-wk1/pointer-lock/pointer-lock-option-unadjusted-movement-expected.txt: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::requestPointerLock):
(WebCore::ChromeClient::requestPointerUnlock):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/PointerLockController.cpp:
(WebCore::PointerLockController::ref const):
(WebCore::PointerLockController::deref const):
(WebCore::PointerLockController::requestPointerLock):
(WebCore::PointerLockController::requestPointerUnlock):
(WebCore::PointerLockController::requestPointerUnlockAndForceCursorVisible):
* Source/WebCore/page/PointerLockController.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestPointerLock):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
(WKPageDidAllowPointerLock):
(WKPageDidDenyPointerLock):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/API/glib/WebKitPointerLockPermissionRequest.cpp:
(webkitPointerLockPermissionRequestAllow):
(webkitPointerLockPermissionRequestDeny):
(webkitPointerLockPermissionRequestCreate):
(webkitPointerLockPermissionRequestDidLosePointerLock):
* Source/WebKit/UIProcess/API/glib/WebKitPointerLockPermissionRequestPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRequestPointerLock):
(webkitWebViewDenyPointerLockRequest):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseRequestPointerLock):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestPointerLock):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::requestPointerLock):
(WebKit::WebPageProxy::didAllowPointerLock):
(WebKit::WebPageProxy::didDenyPointerLock):
(WebKit::WebPageProxy::requestPointerUnlock):
(WebKit::WebPageProxy::webContentPointerLockProcess):
(WebKit::WebPageProxy::clearWebContentPointerLockProcess):
(WebKit::WebPageProxy::resetPointerLockState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setShouldSuppressFirstResponderChanges):
(WebKit::WebViewImpl::windowWillBeginSheet):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::requestPointerLock):
(WebKit::WebChromeClient::requestPointerUnlock):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::requestPointerLock):
(WebChromeClient::requestPointerUnlock):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::requestPointerLock):

Canonical link: <a href="https://commits.webkit.org/297606@main">https://commits.webkit.org/297606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/864c79c7cf88efca14a3d1f7c1a2d984ff7e86b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112287 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22496 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118365 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62637 "An unexpected error occured. Recent messages:Printed configuration") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32671 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40582 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85300 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/62637 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115234 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26067 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101023 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65730 "Passed tests") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/25372 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19159 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62211 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19238 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121691 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39361 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/40582 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94112 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39742 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93936 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39173 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16967 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35392 "Build is in progress. Recent messages:") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18099 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39249 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44737 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38884 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42221 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40627 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -730,20 +730,20 @@ OptionSet<PointerCharacteristics> WebChromeClient::pointerCharacteristicsOfAllAv
 {
     if (RefPtr page = m_page.get())
-        page->send(Messages::WebPageProxy::RequestPointerUnlock());
+        page->sendWithAsyncReply(Messages::WebPageProxy::RequestPointerUnlock(), WTFMove(completion));
```
We need to call the CompletionHandler if there is no page.

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -730,20 +730,20 @@ OptionSet<PointerCharacteristics> WebChromeClient::pointerCharacteristicsOfAllAv
 {
     RefPtr page = m_page.get();
     if (!page)
         return false;
```
We need to call the CompletionHandler if there's no page.  Also, we may be able to remove the return value of this function.

achristensen07 commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2126,7 +2126,7 @@ static void setDynamicRangeLimit(CALayer* layer, PlatformDynamicRangeLimit platf
 {
 #if ENABLE(POINTER_LOCK)
-    m_page->requestPointerUnlock();
+    m_page->requestPointerUnlock(nullptr);
```
This looks like it will crash.  Give it a lambda that does the equivalent of WebPage::DidAcquirePointerLock if successful, something else if unsuccessful.  Same with all the places you pass nullptr.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -1751,15 +1751,15 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
 
     auto checker = CompletionHandlerCallChecker::create(delegate.get(), @selector(_webViewDidRequestPointerLock:completionHandler:));
-    [static_cast<id<WKUIDelegatePrivate>>(delegate) _webViewDidRequestPointerLock:uiDelegate->m_webView.get().get() completionHandler:makeBlockPtr([checker = WTFMove(checker), page = RefPtr { page }] (BOOL allow) {
+    [static_cast<id<WKUIDelegatePrivate>>(delegate) _webViewDidRequestPointerLock:uiDelegate->m_webView.get().get() completionHandler:makeBlockPtr([checker = WTFMove(checker), page = RefPtr { page }, reply = WTFMove(reply)] (BOOL allow) {
```
put mutable here before the last { instead of const_cast below.

achristensen07 commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -142,7 +148,12 @@ void PointerLockController::requestPointerUnlock()
 
     m_unlockPending = true;
-    m_page.chrome().client().requestPointerUnlock();
+    m_page.chrome().client().requestPointerUnlock([&] (bool result) {
```
This unsafely captures a raw pointer.

achristensen07 commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -733,7 +733,7 @@ - (void)setIsSelected:(BOOL)isSelected;
 
 #if ENABLE(POINTER_LOCK)
-bool WebChromeClient::requestPointerLock()
+bool WebChromeClient::requestPointerLock(CompletionHandler<void(bool)>&&)
```
This needs to do something with the CompletionHandler.

jesxilin commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -142,7 +148,12 @@ void PointerLockController::requestPointerUnlock()
 
     m_unlockPending = true;
-    m_page.chrome().client().requestPointerUnlock();
+    m_page.chrome().client().requestPointerUnlock([this] (bool result) {
```
This code is repeated in this file, so I think it could be good to put in a function instead.

achristensen07 replied:
Also unsafe use of pointer capture.  Nothing is guaranteeing this has not been deallocated when the callback happens.

achristensen07 commented on LayoutTests/platform/mac-site-isolation/TestExpectations:
```diff
@@ -9,7 +9,7 @@ webkit.org/b/280402 wktr/notify-done-moment-dump-as-text-onload-async.html [ Fai
 # Root directories skipped until we can make test expectations for them: #
 ##########################################################################
-imported/ [ Skip ]
+# imported/ [ Skip ]
```
Don't make this change

achristensen07 commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -128,7 +128,13 @@ void PointerLockController::requestPointerLock(Element* target, std::optional<Po
         // ChromeClient::requestPointerLock() can call back into didAcquirePointerLock(), so all state including element, options, and promise needs to be stored before it is called.
-        if (!m_page.chrome().client().requestPointerLock()) {
+
+        if (!m_page.chrome().client().requestPointerLock([&, promise = WTFMove(promise)] (bool result) {
```
This unsafely captures the this pointer.

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2232,12 +2232,15 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
                 return;
+            }
             
             m_client.requestPointerLock(toAPI(page), m_client.base.clientInfo);
```
I guess since this is only used for testing this is fine.

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -3020,7 +3023,7 @@ void WKPageDidAllowPointerLock(WKPageRef pageRef)
     CRASH_IF_SUSPENDED;
 #if ENABLE(POINTER_LOCK)
-    toImpl(pageRef)->didAllowPointerLock();
+    toImpl(pageRef)->didAllowPointerLock([](bool) { });
```
We need to remove WKPageDidAllowPointerLock instead.

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -3040,7 +3043,7 @@ void WKPageDidDenyPointerLock(WKPageRef pageRef)
     CRASH_IF_SUSPENDED;
 #if ENABLE(POINTER_LOCK)
-    toImpl(pageRef)->didDenyPointerLock();
+    toImpl(pageRef)->didDenyPointerLock([](bool) { });
```
Remove WKPageDidDenyPointerLock

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -1732,34 +1732,41 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
     RefPtr uiDelegate = m_uiDelegate.get();
-    if (!uiDelegate)
+    if (!uiDelegate) {
+        reply(false);
```
nit: I like "return reply(false);" because it's one line and doesn't require { }.  Some feel otherwise.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14313,37 +14315,40 @@ void WebPageProxy::didAllowPointerLock()
-void WebPageProxy::requestPointerUnlock()
+void WebPageProxy::requestPointerUnlock(CompletionHandler<void(bool)>&& callback)
 {
+    bool isPointerLocked = m_isPointerLocked;
```
I'd call this "wasPointerLocked"

achristensen07 reviewed (changes requested):
This needs test coverage with site isolation on.

cdumez commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -580,8 +580,8 @@ class ChromeClient {
 #if ENABLE(POINTER_LOCK)
-    virtual bool requestPointerLock() { return false; }
-    virtual void requestPointerUnlock() { }
+    virtual bool requestPointerLock(CompletionHandler<void(bool)>&&) { return false; }
```
Not calling the completion handler here is a bug.

cdumez commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -580,8 +580,8 @@ class ChromeClient {
-    virtual bool requestPointerLock() { return false; }
-    virtual void requestPointerUnlock() { }
+    virtual bool requestPointerLock(CompletionHandler<void(bool)>&&) { return false; }
+    virtual void requestPointerUnlock(CompletionHandler<void(bool)>&&) { }
```
Not calling the completion handler here is a bug.

cdumez commented on Source/WebCore/page/Page.h:
```diff
@@ -367,6 +367,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     WEBCORE_EXPORT static Ref<Page> create(PageConfiguration&&);
     WEBCORE_EXPORT ~Page();
 
+
```
Unnecessary change

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -366,7 +366,7 @@ Page::Page(PageConfiguration&& pageConfiguration)
     , m_pointerCaptureController(makeUniqueRef<PointerCaptureController>(*this))
 #if ENABLE(POINTER_LOCK)
-    , m_pointerLockController(makeUniqueRef<PointerLockController>(*this))
+    , m_pointerLockController(makeUniqueWithoutRefCountedCheck<PointerLockController>(*this))
```
`makeUniqueRefWithoutRefCountedCheck`

cdumez commented on Source/WebCore/page/Page.h:
```diff
@@ -498,7 +499,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     PointerCaptureController& pointerCaptureController() { return m_pointerCaptureController.get(); }
 #if ENABLE(POINTER_LOCK)
-    PointerLockController& pointerLockController() { return m_pointerLockController.get(); }
+    PointerLockController* pointerLockController() { return m_pointerLockController.get(); }
```
Change in the wrong direction.

cdumez commented on Source/WebCore/page/Page.h:
```diff
@@ -1441,7 +1442,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     const UniqueRef<PointerCaptureController> m_pointerCaptureController;
 #if ENABLE(POINTER_LOCK)
-    const UniqueRef<PointerLockController> m_pointerLockController;
+    const std::unique_ptr<PointerLockController> m_pointerLockController;
```
Change in the wrong direction.

cdumez commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -153,7 +165,12 @@ void PointerLockController::requestPointerUnlockAndForceCursorVisible()
 
     m_unlockPending = true;
-    m_page.chrome().client().requestPointerUnlock();
+    m_page.chrome().client().requestPointerUnlock([this] (bool result) {
```
Why is it safe to not protected `this` here?

cdumez replied:
nit: `no space between `]` and `(`.

cdumez commented on Source/WebKit/UIProcess/API/APIUIClient.h:
```diff
@@ -190,7 +190,7 @@ class UIClient {
 
 #if ENABLE(POINTER_LOCK)
-    virtual void requestPointerLock(WebKit::WebPageProxy*) { }
+    virtual void requestPointerLock(WebKit::WebPageProxy*, CompletionHandler<void(bool)>&&) { }
```
It is a bug to not call the completion handler here.

cdumez commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -1737,34 +1737,34 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
 #if ENABLE(POINTER_LOCK)
 
-void UIDelegate::UIClient::requestPointerLock(WebPageProxy* page)
+void UIDelegate::UIClient::requestPointerLock(WebPageProxy* page, CompletionHandler<void(bool)>&& reply)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14314,23 +14314,25 @@ void WebPageProxy::closeOverlayedViews()
 }
 
-void WebPageProxy::didAllowPointerLock()
+void WebPageProxy::didAllowPointerLock(CompletionHandler<void(bool)>&& callback)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14314,23 +14314,25 @@ void WebPageProxy::closeOverlayedViews()
 
 #if ENABLE(POINTER_LOCK)
-void WebPageProxy::requestPointerLock()
+void WebPageProxy::requestPointerLock(CompletionHandler<void(bool)>&& callback)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14339,37 +14341,40 @@ void WebPageProxy::didAllowPointerLock()
 }
 
-void WebPageProxy::didDenyPointerLock()
+void WebPageProxy::didDenyPointerLock(CompletionHandler<void(bool)>&& callback)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14339,37 +14341,40 @@ void WebPageProxy::didAllowPointerLock()
 }
 
-void WebPageProxy::requestPointerUnlock()
+void WebPageProxy::requestPointerUnlock(CompletionHandler<void(bool)>&& callback)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -732,20 +732,24 @@ OptionSet<PointerCharacteristics> WebChromeClient::pointerCharacteristicsOfAllAv
 #if ENABLE(POINTER_LOCK)
 
-bool WebChromeClient::requestPointerLock()
+bool WebChromeClient::requestPointerLock(CompletionHandler<void(bool)>&& completion)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -732,20 +732,24 @@ OptionSet<PointerCharacteristics> WebChromeClient::pointerCharacteristicsOfAllAv
 }
 
-void WebChromeClient::requestPointerUnlock()
+void WebChromeClient::requestPointerUnlock(CompletionHandler<void(bool)>&& completion)
```
Can we call this `completionHandler` like we usually do?

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -733,29 +733,29 @@ - (void)setIsSelected:(BOOL)isSelected;
 {
 #if PLATFORM(MAC)
     if (![m_webView page])
         return false;
```
Failure to call the completion handler, this is a bug.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -733,29 +733,29 @@ - (void)setIsSelected:(BOOL)isSelected;
-    [m_webView page]->pointerLockController().didAcquirePointerLock();
+    [m_webView page]->pointerLockController()->didAcquirePointerLock();
     
     return true;
```
Failure to call the completion handler, this is a bug.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -733,29 +733,29 @@ - (void)setIsSelected:(BOOL)isSelected;
     
     return true;
 #else
     return false;
```
Failure to call the completion handler, this is a bug.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -733,29 +733,29 @@ - (void)setIsSelected:(BOOL)isSelected;
 }
 
-void WebChromeClient::requestPointerUnlock()
+void WebChromeClient::requestPointerUnlock(CompletionHandler<void(bool)>&&)
```
Failure to call the completion handler, this is a bug.

cdumez commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -128,7 +129,13 @@ void PointerLockController::requestPointerLock(Element* target, std::optional<Po
         // ChromeClient::requestPointerLock() can call back into didAcquirePointerLock(), so all state including element, options, and promise needs to be stored before it is called.
-        if (!m_page.chrome().client().requestPointerLock()) {
+
+        if (!m_page.chrome().client().requestPointerLock([this, protectedThis = Ref { *this }, promise = WTFMove(promise)] (bool result) {
```
nit: `no space between `]` and `(`.

cdumez commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -142,7 +149,12 @@ void PointerLockController::requestPointerUnlock()
 
     m_unlockPending = true;
-    m_page.chrome().client().requestPointerUnlock();
+    m_page.chrome().client().requestPointerUnlock([this, protectedThis = Ref { *this }] (bool result) {
```
nit: `no space between `]` and `(`.

cdumez commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -1737,34 +1737,34 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
 
     auto checker = CompletionHandlerCallChecker::create(delegate.get(), @selector(_webViewDidRequestPointerLock:completionHandler:));
-    [static_cast<id<WKUIDelegatePrivate>>(delegate) _webViewDidRequestPointerLock:uiDelegate->m_webView.get().get() completionHandler:makeBlockPtr([checker = WTFMove(checker), page = RefPtr { page }] (BOOL allow) {
+    [static_cast<id<WKUIDelegatePrivate>>(delegate) _webViewDidRequestPointerLock:uiDelegate->m_webView.get().get() completionHandler:makeBlockPtr([checker = WTFMove(checker), page = RefPtr { page }, reply = WTFMove(reply)] (BOOL allow) mutable {
```
nit: `no space between `]` and `(`.

rniwa commented on Source/WebCore/page/PointerLockController.h:
```diff
@@ -56,6 +56,9 @@ class PointerLockController {
     void requestPointerLock(Element* target, std::optional<PointerLockOptions>&& = std::nullopt, RefPtr<DeferredPromise> = nullptr);
 
+    void ref() const { m_page.ref(); }
+    void deref() const { m_page.deref(); }
```
I don't think we can do this without including `Page.h`?

stwrt replied:
We define `class page` above. Is that not sufficient?

aprotyas replied:
No, since that's just a declaration, the compiler does not yet know what the `Page` class or its `ref()/deref()` methods are.

darinadler replied:
We can’t do it in the header, but we could do it in the `.cpp` file.

stwrt replied:
Moved the implementation of `ref`/`deref` to the `.cpp` file, and am keeping the function declaration in the header file. Also, added the `Page.h` to the header file.

pascoej reviewed (changes requested):
^

rniwa commented on Source/WebCore/page/PointerLockController.h:
```diff
@@ -27,6 +27,7 @@
 #if ENABLE(POINTER_LOCK)
 
 #include "ExceptionCode.h"
+#include "Page.h"
```
I don't think we should include `Page.h` if we're moving the implementation of ref/deref to cpp file.

darinadler commented on Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/xcshareddata/xcschemes/WebKitTestRunner.xcscheme:
```diff
@@ -71,6 +71,10 @@
             isEnabled = "YES">
          </CommandLineArgument>
+         <CommandLineArgument
+            argument = "/Users/bts/Developer/site_isolation/LayoutTests/pointer-lock/lock-lost-on-alert.html"
```
Don't want to check this in since it's user-specific.

jesxilin commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -580,8 +586,11 @@ class ChromeClient {
-    virtual bool requestPointerLock() { return false; }
-    virtual void requestPointerUnlock() { }
+    virtual void requestPointerLock(CompletionHandler<void(PointerLockRequestResult)>&& completionHandler)
+    {
```
Style nit: lines 590 - 592 should be on the same line as 589, just like how you have it for `virtual void requestPointerUnlock` right below this

stwrt replied:
That's fine, but the line is getting quite long when we do this.

jesxilin commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -62,6 +63,17 @@ PointerLockController::PointerLockController(Page& page)
+{
+    m_page.deref();
+}
+
```
Style Nit: there's double new line (75 & 76)

pxlcoder replied:
This comment was never addressed.

jesxilin commented on Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/xcshareddata/xcschemes/WebKitTestRunner.xcscheme:
```diff
@@ -71,6 +71,10 @@
             argument = "--no-timeout"
             isEnabled = "YES">
          </CommandLineArgument>
+         <CommandLineArgument
```
Please undo changes to this file

jesxilin commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -128,11 +140,22 @@ void PointerLockController::requestPointerLock(Element* target, std::optional<Po
-            clearElement();
-        }
+
+        m_page.chrome().client().requestPointerLock([this, protectedThis = Ref { *this }, promise = WTFMove(promise), target = WTFMove(target)](PointerLockRequestResult result) {
```
`target = WTFMove(target)` should be `target = Ref { *target }` b/c `Element* target`

achristensen07 commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -71,7 +83,7 @@ void PointerLockController::requestPointerLock(Element* target, std::optional<Po
     }
 
-    if (m_documentAllowedToRelockWithoutUserGesture != &target->document() && !UserGestureIndicator::processingUserGesture()) {
+    if (m_documentAllowedToRelockWithoutUserGesture != &target->document() && !UserGestureIndicator::processingUserGesture() && promise) {
```
This change doesn't look good.

stwrt replied:
okay

achristensen07 replied:
I still don't understand why this change was made, and I think we should revert it unless there is a reason to keep it.  If there's a reason to keep it, please describe that reason in the change log.

achristensen07 commented on Source/WebCore/page/PointerLockController.h:
```diff
@@ -26,6 +26,7 @@
 
 #if ENABLE(POINTER_LOCK)
 
+#include "ChromeClient.h"
```
This does not need to be included.

achristensen07 commented on Source/WebKit/UIProcess/API/APIUIClient.h:
```diff
@@ -30,6 +30,7 @@
 #include "WKPage.h"
 #include "WebEvent.h"
 #include "WebHitTestResultData.h"
+#include <WebCore/ChromeClient.h>
```
This should not need to be included.

stwrt replied:
This is needed.

whsieh replied:
I think @achristensen07 is suggesting that you forward-declare `PointerLockRequestResult`, so you don't need this import:

```suggestion
```
…
```
namespace WebCore {
…
enum class PointerLockRequestResult : uint8_t;
…
}
```

achristensen07 commented on Source/WebKit/UIProcess/API/APIUIClient.h:
```diff
@@ -191,7 +192,7 @@ class UIClient {
 
 #if ENABLE(POINTER_LOCK)
-    virtual void requestPointerLock(WebKit::WebPageProxy*) { }
+    virtual void requestPointerLock(WebKit::WebPageProxy*, CompletionHandler<void(WebCore::PointerLockRequestResult)>&& completionHandler) { completionHandler(WebCore::PointerLockRequestResult::Failure); }
```
This should just be a bool.  PointerLockRequestResult is only needed for ChomeClients.

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2240,12 +2241,13 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
 
 #if ENABLE(POINTER_LOCK)
-        void requestPointerLock(WebPageProxy* page) final
+        void requestPointerLock(WebPageProxy* page, CompletionHandler<void(WebCore::PointerLockRequestResult)>&& completionHandler) final
```
bool

achristensen07 commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2122,7 +2122,13 @@ static NSTrackingAreaOptions flagsChangedEventMonitorTrackingAreaOptions()
-    m_page->requestPointerUnlock();
+    m_page->requestPointerUnlock([this](bool result) {
+        if (result) {
+            WeakPtr webContentPointerLockProcess = m_page->webContentPointerLockProcess();
```
This needs to be a RefPtr.
Also, instead of webContentPointerLockProcess and clearWebContentPointerLockProcess, why not just make one function takePointerLockProcess?

Also, this is going to be a 4th copy of the same thing, so it should call the shared new member function of WebPageProxy, too.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3195,7 +3195,7 @@ void WebPageProxy::dispatchActivityStateChange()
     if (((changed & ActivityState::IsVisible) && !isViewVisible()) || ((changed & ActivityState::WindowIsActive) && !protectedPageClient()->isViewWindowActive())
         || ((changed & ActivityState::IsFocused) && !isViewFocused()))
-        requestPointerUnlock();
+        requestPointerUnlock([](bool) { });
```
I assume all these empty lambdas should probably do the same thing as WebViewImpl::windowWillBeginSheet does, unless there is a clear reason why they don't need to.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14387,62 +14387,83 @@ void WebPageProxy::closeOverlayedViews()
+
+    m_uiClient->requestPointerLock(this, [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), webContentProcess = WTFMove(webContentProcess)] (WebCore::PointerLockRequestResult result) mutable {
+        switch (result) {
+        case WebCore::PointerLockRequestResult::Success:
```
UI process should only use bool success/failure instead of PointerLockRequestResult.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -29,6 +29,7 @@
 // Use forward declarations and WebPageProxyInternals.h instead.
 #include "APIObject.h"
 #include "MessageReceiver.h"
+#include <WebCore/ChromeClient.h>
```
Don't include this here.

achristensen07 commented on LayoutTests/http/tests/site-isolation/pointer-lock.html:
```diff
@@ -0,0 +1,39 @@
+<head>
+    <script src="/js-test-resources/js-test.js"></script>
+    <script src="/js-test-resources/ui-helper.js"></script>
+    <meta name="fuzzy" content="maxDifference=112; totalPixels=900" />
```
This should be removed.

achristensen07 commented on LayoutTests/http/tests/site-isolation/pointer-lock.html:
```diff
@@ -0,0 +1,39 @@
+    // Pointer lock not supported on iOS
+    const hasPointerLock = 'pointerLockElement' in document;
+    if (!hasPointerLock) {
+        console.log("Pointerlock not supported");
```
This looks like it'll need iOS-specific test expectations in LayoutTests/platform.

stwrt replied:
Exactly, the test will timeout on iOS since we do not support the pointer lock api.

stwrt replied:
Fixed

achristensen07 commented on LayoutTests/http/tests/site-isolation/resources/pointer-lock.html:
```diff
@@ -0,0 +1,41 @@
+        if (window.eventSender) {
+            UIHelper.activateElement(element);
+        }
+    }, 100);
```
This timeout seems a little short, and short timeouts cause flaky tests.  Can it be removed, or can we instead wait for something to happen in a more direct way?

stwrt replied:
Waiting for `onload` would be better here.

achristensen07 commented on Source/WebCore/page/PointerLockController.h:
```diff
@@ -51,11 +51,15 @@ class WeakPtrImplWithEventTargetData;
 class PointerLockController {
     WTF_MAKE_NONCOPYABLE(PointerLockController);
     WTF_MAKE_TZONE_ALLOCATED(PointerLockController);
+
```
Let's not add a newline here.

achristensen07 commented on Source/WebKit/UIProcess/API/APIUIClient.h:
```diff
@@ -63,6 +63,7 @@ class SecurityOriginData;
 enum class AutoplayEvent : uint8_t;
 enum class AutoplayEventFlags : uint8_t;
 enum class MediaProducerMediaState : uint32_t;
+enum class PointerLockRequestResult : uint8_t;
```
This should be removed.

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -97,6 +97,7 @@
 #include "WebProtectionSpace.h"
 #include <WebCore/AuthenticatorAssertionResponse.h>
 #include <WebCore/AutoplayEvent.h>
+#include <WebCore/ChromeClient.h>
```
This should be removed.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
```diff
@@ -29,6 +29,7 @@
 
 #import "APIContextMenuClient.h"
 #import "APIUIClient.h"
+#import <WebCore/ChromeClient.h>
```
This should be removed.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3195,7 +3196,12 @@ void WebPageProxy::dispatchActivityStateChange()
-        requestPointerUnlock();
+        requestPointerUnlock([this, protectedThis = Ref { *this }](bool result) {
+            if (result) {
+                webContentPointerLockProcess()->send(Messages::WebPage::DidLosePointerLock(), webPageIDInProcess(*webContentPointerLockProcess().get()));
```
Let's capture the result of webContentPointerLockProcess in a RefPtr and null check it.  No need to crash if state gets messed up.

stwrt replied:
Added a null check

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7235,7 +7241,7 @@ void WebPageProxy::didCommitLoadForFrame(IPC::Connection& connection, FrameIdent
         m_pluginMaxZoomFactor = { };
 #if ENABLE(POINTER_LOCK)
-        requestPointerUnlock();
+        requestPointerUnlock([](bool) { });
```
Is there a reason we don't do here what we do in the completion handler in WebPageProxy::dispatchActivityStateChange?  Same in WebPageProxy::resetState.  If we need to do the same thing in all 3 places, the code should be shared.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2125,7 +2126,13 @@ static NSTrackingAreaOptions flagsChangedEventMonitorTrackingAreaOptions()
 {
 #if ENABLE(POINTER_LOCK)
-    m_page->requestPointerUnlock();
+    m_page->requestPointerUnlock([this](bool result) {
```
Please capture a smart pointer here.

pxlcoder commented on LayoutTests/http/tests/site-isolation/pointer-lock.html:
```diff
@@ -0,0 +1,39 @@
+    const hasPointerLock = 'pointerLockElement' in document;
+    if (!hasPointerLock) {
+        console.log("Pointerlock not supported");
+        testRunner.notifyDone();
```
Should probably check `if (window.testRunner)` here for consistency with the other places.

stwrt replied:
Ah, right missed that one.

darinadler commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -127,12 +139,23 @@ void PointerLockController::requestPointerLock(Element* target, std::optional<Po
-        }
+
+        // ChromeClient::requestPointerLock() can call back into didAcquirePointerLock(), so all state including element, options, and target needs to be stored before it is called.
+        m_page.chrome().client().requestPointerLock([this, protectedThis = Ref { *this }, target = WTFMove(target)](PointerLockRequestResult result) {
```
This is doing `WTFMove` on an `Element*`. The move has no effect, but also that’s not right for object lifetime. We need a `Ref<Element>` or `RefPtr<Element>`.

darinadler commented on Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
```diff
@@ -136,7 +136,7 @@ typedef void (*WKCheckUserMediaPermissionCallback)(WKPageRef page, WKFrameRef fr
 typedef void (*WKHandleAutoplayEventCallback)(WKPageRef page, WKAutoplayEvent event, WKAutoplayEventFlags flags, const void* clientInfo);
 typedef void (*WKFullscreenMayReturnToInlineCallback)(WKPageRef page, const void* clientInfo);
-typedef void (*WKRequestPointerLockCallback)(WKPageRef page, const void* clientInfo);
+typedef void (*WKRequestPointerLockCallback)(WKPageRef page, WKCompletionListenerRef listener, const void* clientInfo);
```
This is API on some platforms, so it can’t just have its type changed. We need a versioned change like all the `_deprecatedForUseWithV` things below, and the new version with the listener needs to be in a new `WKPageUIClientV20`.

achristensen07 replied:
I think I disagree.  This interface was designed to be able to support binary compatibility with a changing and expanding API surface, but ever since Safari moved to WKWebView this interface's only client is WebKitTestRunner, which can handle changes like this.  I've verified there are no Apple clients of this callback, and the glib clients maintain their own separate interface.  There are no contributing users that this will break.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3195,7 +3196,16 @@ void WebPageProxy::dispatchActivityStateChange()
-        requestPointerUnlock();
+        requestPointerUnlock([this, protectedThis = Ref { *this }](bool result) {
+            if (result) {
+                WeakPtr<WebProcessProxy> webContentPointerLock = webContentPointerLockProcess();
```
safer-cpp requires having a strong reference to a WebProcessProxy before calling something like WebProcessProxy::send.  Changing this from a WeakPtr to a RefPtr would fix that.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7235,7 +7245,17 @@ void WebPageProxy::didCommitLoadForFrame(IPC::Connection& connection, FrameIdent
+                return;
+
+            if (result) {
+                WeakPtr webContentPointerLock = webContentPointerLockProcess();
```
This needs to be a RefPtr, and we should null check it and do nothing if it is null instead of crashing.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11563,7 +11583,7 @@ void WebPageProxy::resetState(ResetStateReason resetStateReason)
 
 #if ENABLE(POINTER_LOCK)
-    requestPointerUnlock();
+    requestPointerUnlock([](bool) { });
```
Is there a good reason to do nothing here, or should we do the same as in WebPageProxy::didCommitLoadForFrame and WebPageProxy::dispatchActivityStateChange?  I think the latter, and I think that code should be moved to a new function rather than writing the same thing three times.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2771,6 +2769,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     Ref<AboutSchemeHandler> protectedAboutSchemeHandler();
 
+#if ENABLE(POINTER_LOCK)
+    WeakPtr<WebProcessProxy> webContentPointerLockProcess() { return m_webContentPointerLockProcess; }
```
Since all callers need to get a RefPtr instead of a WeakPtr, let's change the return type to RefPtr.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2813,7 +2816,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+    void didDenyPointerLock(CompletionHandler<void(bool)>&&);
+    void requestPointerLock(IPC::Connection&, CompletionHandler<void(bool)>&&);
+
+    WeakPtr<WebProcessProxy> m_webContentPointerLockProcess;
```
This should go with the member variables below instead of with the functions.  Ideally with the other pointer lock related member variables, like m_isPointerLocked.

achristensen07 commented on Tools/SwiftBrowser/SwiftBrowser.xcodeproj/xcshareddata/xcschemes/SwiftBrowser.xcscheme:
```diff
@@ -39,8 +39,7 @@
       debugServiceExtension = "internal"
-      allowLocationSimulation = "YES"
-      internalIOSLaunchStyle = "2">
+      allowLocationSimulation = "YES">
```
This change should not be included with this PR.

achristensen07 commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -742,9 +742,9 @@
-		A176D24B2D6D2A4E00C63915 /* GPUExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2472D6D295400C63915 /* GPUExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
-		A176D24C2D6D2A4E00C63915 /* NetworkingExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2482D6D295400C63915 /* NetworkingExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
-		A176D24D2D6D2A4E00C63915 /* WebContentExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2492D6D295400C63915 /* WebContentExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24B2D6D2A4E00C63915 /* GPUExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2472D6D295400C63915 /* GPUExtension.appex */; platformFilter = ios; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
```
This change should not be included in this PR.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -29,6 +29,7 @@
 // Use forward declarations and WebPageProxyInternals.h instead.
 #include "APIObject.h"
 #include "MessageReceiver.h"
+#include "WebProcessProxy.h"
```
This should stay a forward declaration.  If a function doesn't compile, move it to WebPageProxy.cpp.

stwrt committed:
[Site Isolation] Pointer Lock
https://bugs.webkit.org/show_bug.cgi?id=293767
rdar://117910429

Reviewed by Alex Christensen.

Implement Pointer Lock to work under site isolation.

WK1 change: When disabling pointer lock options a pointerlockerror event will be thrown now
in place of pointerlockchange.

* LayoutTests/http/tests/site-isolation/pointer-lock-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/pointer-lock.html: Added.
* LayoutTests/http/tests/site-isolation/resources/pointer-lock.html: Added.
* LayoutTests/platform/ios/http/tests/site-isolation/pointer-lock-expected.txt: Added.
* LayoutTests/platform/mac-wk1/pointer-lock/pointer-lock-option-unadjusted-movement-expected.txt: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::requestPointerLock):
(WebCore::ChromeClient::requestPointerUnlock):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/PointerLockController.cpp:
(WebCore::PointerLockController::ref const):
(WebCore::PointerLockController::deref const):
(WebCore::PointerLockController::requestPointerLock):
(WebCore::PointerLockController::requestPointerUnlock):
(WebCore::PointerLockController::requestPointerUnlockAndForceCursorVisible):
* Source/WebCore/page/PointerLockController.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestPointerLock):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
(WKPageDidAllowPointerLock):
(WKPageDidDenyPointerLock):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/API/glib/WebKitPointerLockPermissionRequest.cpp:
(webkitPointerLockPermissionRequestAllow):
(webkitPointerLockPermissionRequestDeny):
(webkitPointerLockPermissionRequestCreate):
(webkitPointerLockPermissionRequestDidLosePointerLock):
* Source/WebKit/UIProcess/API/glib/WebKitPointerLockPermissionRequestPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRequestPointerLock):
(webkitWebViewDenyPointerLockRequest):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseRequestPointerLock):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestPointerLock):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::requestPointerLock):
(WebKit::WebPageProxy::didAllowPointerLock):
(WebKit::WebPageProxy::didDenyPointerLock):
(WebKit::WebPageProxy::requestPointerUnlock):
(WebKit::WebPageProxy::webContentPointerLockProcess):
(WebKit::WebPageProxy::clearWebContentPointerLockProcess):
(WebKit::WebPageProxy::resetPointerLockState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setShouldSuppressFirstResponderChanges):
(WebKit::WebViewImpl::windowWillBeginSheet):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::requestPointerLock):
(WebKit::WebChromeClient::requestPointerUnlock):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::requestPointerLock):
(WebChromeClient::requestPointerUnlock):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::requestPointerLock):

Canonical link: https://commits.webkit.org/297606@main
